### ISSUE TITLE
SDL2 deadlock with callback scheduler

### ISSUE THREAD

**Issue by mathiasbredholt** [pure-data/pure-data#2297]
Hi,

I'm writing an external that uses SDL2 to create a window in Pd. I managed to create a window and I'm refreshing the window using a clock. All is working fine except that I can't poll user events from the window when DSP is on (callback scheduler active?). When I activate DSP, Pd goes to a deadlock it seems.

It works fine to poll window events (`SDL_PollEvent`) when DSP is inactive (polling scheduler active?). If I don't poll window events and just refresh the screen (`SDL_GL_SwapWindow`) things are working fine aswell.

I'm on latest master branch (after scheduler improvement merge, which did allow me to toggle DSP while window was open, huge thanks!!)

So in conclusion, I can toggle DSP with my window if I don't poll user events. If I toggle DSP while polling user events, Pd freezes.

Any hints to where to look for a solution?

Thanks in advance

**Comment by Spacechild1** [comment]
Hi, I had to go down this exact rabbit hole when I developed `[vstplugin~]`.

Short answer: there is currently no reliable and cross-platform way for managing custom windows from within Pd.

Long answer: SDL2 provides a nice abstraction layer for window management; however, some OS have peculiar requirements that will always "leak" from the API.

Typically, there is a single (input) event loop that is managed by the application. Unfortunately, Pd does not have an event loop, so externals must somehow try to create their own. Let's look at the main big OS:

* Windows: windows and its associated message queues are owned by the thread that created them. This means that you must create windows on the same thread that calls `GetMessage`!

* macOS: an application may only have a single UI event loop which *must* run on the main thread!

* Linux: event loops are associated with displays (`Display`); it is therefore possible to create windows and call `XNextEvent` on different threads, but it's probably not recommended.

Pd's polling scheduler runs on the main thread, but the callback scheduler runs on the audio callback. If DSP is turned off, Pd always switches to the polling scheduler. If you try to poll input events with a clock in the callback scheduler, you will run into issues (e.g. deadlocks) on Windows and macOS if your window has been created in the polling scheduler, i.e. on the "wrong" thread.

For Windows the solution is to run the UI event loop in a separate thread and communicate with Pd's scheduler via a threadsafe queue(s). If you want to receive messages from the UI thread, you need to regularly poll the queue with a clock. *Do not try to use `sys_lock`/`sys_unlock` because this can cause deadlocks!*

A nice side effect of having a dedicated UI thread is that UI updates do not disturb audio processing, so I would also recommend doing the same on Linux.

On macOS you have no choice but poll events from the polling scheduler thread and tell the user *not* to use the callback scheduler. Side note: this only happens to work because Pd runs the polling scheduler on the main thread by default; otherwise you would be completely out of luck.

---

FWIW, I made a PR to add a Cocoa event loop to Pd, so we wouldn't need to poll the UI on the audio thread: https://github.com/pure-data/pure-data/pull/1117.

I think I should go ahead and also implement event loops for Windows and Linux and add API functions for externals to exchange messages between the UI thread and the scheduler thread. It could look like this:
```
/* send a message to the UI thread; if 'sync' is 1, block until the message has been dispatched */
EXTERN void sys_eventloop_send(void *data, t_sendfn fn, int sync);

/* send a message to a Pd object from another thread */
EXTERN void pd_queue_mess(t_pdinstance *instance, t_pd *obj, void *data, t_sendfn fn);
/* cancel all pending thread messages for a Pd object */
EXTERN void pd_queue_cancel(t_pd *obj);
```

`pd_queue_mess` would be useful for many other scenarios where people want to send messages from helper/worker threads back to Pd. (Again, you cannot use `sys_lock` and `sys_unlock` because it can deadlock.)

In the meantime, you'd need to implement everything manually (as outlined above)...

Anyway, thanks for opening this issue because it shows that I'm not the only one who'd need this functionality.

**Comment by sebshader** [comment]
what is the ```sys_idlehook()``` for? the comments say it can be plugged into but it's a bit odd that there's only 1 'slot'.
seems like it might be a reasonable place to run gui updates?

**Comment by Spacechild1** [comment]
> seems like it might be a reasonable place to run gui updates?

The idlehook is called on the scheduler thread, so it wouldn't solve anything. As I tried to explain above, on macOS/Windows it isn't even possible to poll the UI from the callback scheduler.

> edit: I see in the event loop PR that it runs the audio on another thread afaict, that's probably better..

Exactly! We certainly want to run the UI and the scheduler in different threads.

**Comment by mathiasbredholt** [comment]
@Spacechild1 hi thanks a lot for your response, I'm on macOS so yeah I can't do polling/updates from separate thread. 

I tried your PR and I can't figure out how to hook into the event loop. Where should I place `SDL_PollEvent` and  `SDL_GL_SwapWindow`? 

I hope your PR gets merged, as it is not very nice at all to render the graphics from the audio callback.

**Comment by Spacechild1** [comment]
> Where should I place SDL_PollEvent and SDL_GL_SwapWindow?

You wouldn't call `SDL_PollEvent` at all because the plugin does not own the event loop! Instead you would set event handlers for *your* window. After all, your plugin might not be the only one! Don't know how you would do this with SDL, though.

In general, I think SDL is meant for game-loop style applications and expects that you have control over the whole application. I think it's not really suited for plugin UIs, you'd rather need something like JUCE - or write your own little window abstraction (as I did for `vstplugin~`).

Regarding `SDL_GL_SwapWindow`: if you want periodically draw things on a typcal *blocking* event loop, you'd need to install a timer. However, you can also do the OpenGL draw calls in your own thread.

---

If your external is supposed to "take over" Pd - i.e. it does not need to cooperate with other plugins that also create native windows - you can poll the input events from the main Pd thread and do all your OpenGL drawing in a separate thread. I think that's the approach chosen by https://github.com/Ant1r/ofxPof. 

However, this only works as long as Pd does not run in the callback scheduler. At least you could prevent the deadlock by checking if you are on the correct thread. On macOS you can do `[NSThread isMainThread]`; on Windows you can save the thread ID when you create the window and then compare it with the current thread. If you are on the "wrong" thread, you would skip `SDL_PollEvent` and post a warning instead.

**Comment by Spacechild1** [comment]
Have you considered moving the graphics into a separate process? This way you could just write an ordinary SDL application that communicates with your Pd external via sockets/pipes/shared memory, etc.

**Comment by mathiasbredholt** [comment]
@Spacechild1 thanks again for your elaborate answers!

As I need to always be able to poll the window (otherwise it won't open or close) I should probably move the SDL window into a separate process. Definitely easier to get working and maintain for different platforms.

Then my external could spawn the process with `popen` and communicate through a pipe. The binary could be placed next to the external and then maybe use `[ file which ]` to resolve the path to the external.

I still think it would be nice if it was possible to hook into the main loop from externals without being forced to use the callback scheduler, but I guess this is what we have for now.
